src/main/java/org/dcsa/tnt/service/impl/TNTEventSubscriptionTOServiceImpl.java

```diff
@@ -0,0 +1,347 @@
+    extends EventSubscriptionTOServiceImpl<TNTEventSubscriptionTO, EventSubscriptionService> {
+
+  private static final List<EventType> ALL_ALLOWED_EVENT_TYPES =
+      List.of(EventType.SHIPMENT, EventType.TRANSPORT, EventType.EQUIPMENT);
+
+  private final EventSubscriptionService eventSubscriptionService;
+  private final EventSubscriptionRepository eventSubscriptionRepository;
+
+  @Override
+  protected EventSubscriptionService getService() {
+    return this.eventSubscriptionService;
+  }
+
+  @Override
+  public Mono<TNTEventSubscriptionTO> create(TNTEventSubscriptionTO eventSubscriptionTO) {
+    return eventSubscriptionService
+        .create(eventSubscriptionTOToEventSubscription.apply(eventSubscriptionTO))
+        .flatMap(
+            eventSubscription -> {
+              eventSubscriptionTO.setSubscriptionID(eventSubscription.getSubscriptionID());
+              return createEventTypes(eventSubscriptionTO)
+                  .then(createTransportDocumentEventTypes(eventSubscriptionTO))
+                  .then(createShipmentEventType(eventSubscriptionTO))
+                  .then(createTransportEventType(eventSubscriptionTO))
+                  .then(createEquipmentEventType(eventSubscriptionTO))
+                  .thenReturn(eventSubscriptionTO);
+            });
+  }
+
+  // ToDo : replace this with mapstruct
+  private final Function<TNTEventSubscriptionTO, EventSubscription>
+      eventSubscriptionTOToEventSubscription =
+          esTo -> {
+            EventSubscription eventSubscription = new EventSubscription();
+            eventSubscription.setSubscriptionID(esTo.getSubscriptionID());
+            eventSubscription.setCallbackUrl(esTo.getCallbackUrl());
+            eventSubscription.setSecret(esTo.getSecret());
+            eventSubscription.setCarrierBookingReference(esTo.getCarrierBookingReference());
+            eventSubscription.setEquipmentReference(esTo.getEquipmentReference());
+            eventSubscription.setCarrierServiceCode(esTo.getCarrierServiceCode());
+            eventSubscription.setCarrierVoyageNumber(esTo.getCarrierVoyageNumber());
+            eventSubscription.setVesselIMONumber(esTo.getVesselIMONumber());
+            eventSubscription.setTransportDocumentReference(esTo.getTransportDocumentReference());
+            eventSubscription.setTransportCallID(esTo.getTransportCallID());
+            return eventSubscription;
+          };
+
+  @Override
+  public Mono<TNTEventSubscriptionTO> update(TNTEventSubscriptionTO eventSubscriptionTO) {
+    if (eventSubscriptionTO.getSecret() != null) {
+      return Mono.error(
+          new UpdateException(
+              "Please omit the \"secret\" attribute.  If you want to change the"
+                  + " secret, please use the dedicated secret endpoint"
+                  + " (\"PUT .../event-subscriptions/"
+                  + eventSubscriptionTO.getSubscriptionID()
+                  + "/secret\")."));
+    }
```

Comment:
Could this be done in the subclass or a preUpdateHook instead, so we do not have to repeat it in every implementation?